### DISCUSSION TITLE
Getting position by TDoA

### DISCUSSION THREAD

**Discussion by SergeyLadanov** [bitcraze/discussions#357] [Q&A] [answered]
Hi!
I found two functions in the project [crazyflie-firmware](https://github.com/bitcraze/crazyflie-firmware) for processing TDoA data, these are "kalmanCoreUpdateWithTDOA" (file mm_tdoa.c) and "kalmanCoreRobustUpdateWithTDOA" (file mm_tdoa_robust.c).
Which is better? In which cases should I use each of them?

**Comment by krichardsson** [comment, marked as answer]
Hi!

The two functions are called as part of the kalman filter when using the Loco system in TDoA mode, they should usually not be called by external code. `kalmanCoreUpdateWithTDOA()` is used if the parameter `kalman.robustTdoa`is 0, while `kalmanCoreRobustUpdateWithTDOA()` is used when the parameter is non zero.

> Which is better? In which cases should I use each of them?

In my experience the robust flavor is less noisy and more robust(!) but does not converge as easy when the estimated position is off. A middle ground is maybe to let the position first converge using the normal mode and then set the parameter to 1 to switch to the robust mode.

The robust mode also uses a bit more CPU power.

**Reply by SergeyLadanov** [reply]
Hi!
Thanks a lot!
